tensorflow/lite/micro/kernels/xtensa_hifi/pooling.cc

```diff
@@ -0,0 +1,662 @@
+    p_scratch = (void *)xtensa_nnlib_scratch_buf;
+
+    required_scratch = xa_nn_avgpool_getsize(
+            depth,
+            inp_precision,
+            out_precision,
+            input_height,
+            input_width,
+            kernel_height,
+            kernel_width,
+            stride_width, //x_stride,
+            stride_height, //y_stride,
+            pad_width, //x_padding,
+            pad_height, //y_padding,
+            output_height,
+            output_width,
+            inp_data_format,
+            out_data_format);
+
+    if(required_scratch <= 0){
+        TF_LITE_KERNEL_LOG(
+                context, "AveragepoolFloat: xa_nn_avgpool_getsize failed");
+        return kTfLiteError;
+    }
+
+    if(required_scratch > (int)XTENSA_NNLIB_MAX_SCRATCH_SIZE){
+        TF_LITE_KERNEL_LOG(
+                context, "AveragepoolFloat: insufficient scratch memory");
+        return kTfLiteError;
+    }
+
+    inp_data_ptr = GetTensorData<float>(input);
+    out_data_ptr = GetTensorData<float>(output);
+
+    for (int batch = 0; batch < batches; ++batch) {
+        err =  xa_nn_avgpool_f32(&out_data_ptr[output_height*output_width*depth*batch],
+                &inp_data_ptr[output_height*output_width*depth*batch],
+                input_height,
+                input_width,
+                depth,
+                kernel_height,
+                kernel_width,
+                stride_width,
+                stride_height,
+                pad_width,
+                pad_height,
+                output_height,
+                output_width,
+                inp_data_format,
+                out_data_format,
+                p_scratch);
+
+        CHECK_ERR_HIFI_NNLIB_KER(err, "AveragepoolFloat: xa_nn_avgpool_f32 failed");
+    }
+
+    out_length = batches*output_height*output_width*depth;
+    uint32 p_unalign_val = (uint32)out_data_ptr, p_align_val;
+    p_align_val = (p_unalign_val + 7) & (~7);
+
+    //pre loop for activation_min_max
+    int pre_loop_count = p_align_val - p_unalign_val;
+    pre_loop_count = MIN(pre_loop_count, out_length);
+
+    for(int i = 0; i < pre_loop_count; i++) {
+        ACTIVATION_MIN_MAX(float, out_data_ptr[i],
+                out_data_ptr[i], activation_min, activation_max)
+    }
+
+    out_length = out_length - pre_loop_count;
+
+    if(out_length){
+        err = xa_nn_vec_activation_min_max_f32_f32(
+                out_data_ptr,
+                out_data_ptr,
+                activation_min,
+                activation_max,
+                out_length);
+
+        CHECK_ERR_HIFI_NNLIB_KER(err, "AveragepoolFloat: xa_nn_vec_activation_min_max_f32_f32 failed");
+    }
+  return kTfLiteOk;
+}
+
+TfLiteStatus AverageEvalQuantized(TfLiteContext* context, const TfLiteNode* node,
+                      const TfLitePoolParams* params, const OpData* data,
+                      const TfLiteTensor* input, TfLiteTensor* output) {
+  assert(input->type == kTfLiteUInt8 || input->type == kTfLiteInt8);
```

Comment:
`assert is not allowed in TFLM, please use TFLITE_DCHECK instead`

Reply:
@rthadur uploaded new changes